include/sundials/sundials_errors.h

```diff
@@ -0,0 +1,198 @@
+  SUN_SUCCESS                                           = 0
+};
+
+/* ----------------------------------------------------------------------------
+ * Error handler definitions
+ * ---------------------------------------------------------------------------*/
+
+#ifdef __cplusplus /* wrapper to enable C++ usage */
+extern "C" {
+#endif
+
+typedef int (*SUNErrHandlerFn)(int line, const char* func, const char* file,
+                               const char* msg, SUNErrCode err_code,
+                               void* err_user_data, SUNContext sunctx);
+
+struct SUNErrHandler_
+{
+  struct SUNErrHandler_* previous; /* next error handler to call (singly
+                                      linked-list) */
+  SUNErrHandlerFn call;
+  void* data;
+};
+
+typedef struct SUNErrHandler_* SUNErrHandler;
+
+SUNDIALS_EXPORT
+SUNErrHandler SUNErrHandler_Create(SUNErrHandlerFn eh_fn, void* eh_data);
+
+SUNDIALS_EXPORT
+void SUNErrHandler_Destroy(SUNErrHandler eh);
+
+SUNDIALS_EXPORT
+int SUNLogErrHandlerFn(int line, const char* func, const char* file,
+                       const char* msg, SUNErrCode err_code,
+                       void* err_user_data, SUNContext sunctx);
+
+SUNDIALS_EXPORT
+int SUNAbortErrHandlerFn(int line, const char* func, const char* file,
+                         const char* msg, SUNErrCode err_code,
+                         void* err_user_data, SUNContext sunctx);
+
+SUNDIALS_EXPORT
+int SUNAssertErrHandlerFn(int line, const char* func, const char* file,
+                          const char* msg, SUNErrCode err_code,
+                          void* err_user_data, SUNContext sunctx);
+
+/* ----------------------------------------------------------------------------
+ * Error functions
+ * ---------------------------------------------------------------------------*/
+
+/* Turn error code into error message */
+SUNDIALS_EXPORT
+const char* SUNGetErrMsg(SUNErrCode code, SUNContext sunctx);
+
+/* Alternative function to SUNContext_GetLastError that is more concise. */
+static inline SUNErrCode SUNGetLastErr(SUNContext sunctx)
+{
+  SUNErrCode code = SUN_SUCCESS;
+  (void)SUNContext_GetLastError(sunctx, &code);
+  return code;
+}
+
+/* Alternative function to SUNContext_SetLastError that is more concise. */
+static inline SUNErrCode SUNSetLastErr(SUNErrCode code, SUNContext sunctx)
+{
+  sunctx->last_err = code;
+  return SUN_SUCCESS;
+}
+
+/* Alternative function to SUNContext_PeekLastError that is more concise. */
+static inline SUNErrCode SUNPeekLastErr(SUNContext sunctx)
+{
+  SUNErrCode code = SUN_SUCCESS;
+  (void)SUNContext_PeekLastError(sunctx, &code);
+  return code;
+}
+
+static inline void SUNHandleErr(int line, const char* func, const char* file,
+                                SUNErrCode code, SUNContext sunctx)
+{
+  sunctx->last_err = code;
+  SUNErrHandler eh = sunctx->err_handler;
+  while (eh != NULL)
+  {
+    eh->call(line, func, file, NULL, code, eh->data, sunctx);
```

Comment:
I really like the idea of using a linked list for error handlers, and then calling them in sequence so that all have a turn.  This could allow us to have module-specific error handlers lined up, each treating only the errors that can arise within that module (so long as the `SUNErrCode` values are distinct).

Reply:
I cant say this was a unique idea, but I like it. This is how petsc does it too.